### PR TITLE
Add LSTM support [WIP]

### PR DESCRIPTION
**Pull request by LaRiffle**
Address #1868

I'm not very happy with this PR but it's ready for review

### CODE REVIEW

**Comment by robert-wagner** [comment]
What do you not like about this pr?

File: syft/frameworks/torch/hook.py [hunk 1]
```diff
@@ -991,7 +1011,31 @@ def module_float_precision_(nn_self):
 
         self.torch.nn.Module.float_precision = module_float_precision_
 
-        def module_copy_(nn_self):
+        def module_copy(nn_self):
+            """Returns a copy of a torch.nn.Module"""
```
[the remark this answers is not in the document]
Comment (mari-linhares) [reply syft/frameworks/torch/hook.py]: *missing period.

File: syft/serde.py [hunk 2]
```diff
@@ -213,7 +213,9 @@ def numpy_tensor_deserializer(tensor_bin) -> torch.Tensor:
 def torch_tensor_serializer(tensor) -> bin:
     """Strategy to serialize a tensor using Torch saver"""
     binary_stream = io.BytesIO()
+    tensor.flip_hook_native_size()
```
Comment (LaRiffle) [inline syft/serde.py]: @robert-wagner This is what I don't like here

File: syft/frameworks/torch/hook.py [hunk 3]
```diff
@@ -810,6 +810,26 @@ def is_wrapper(self, it_is_a_wrapper):
         tensor_type.native_shape = tensor_type.shape
         tensor_type.native_data = tensor_type.data
 
+        def size(self, dim=None):
+            """Hook the size to return the shape with a callable syntax"""
```
[the remark this answers is not in the document]
Comment (DanyEle) [reply syft/frameworks/torch/hook.py]: I'll to answer to you @mari-linhares , as I just tested this code and it all seems to be working. Basically the `dim` argument is used to specify which dimension of the Tensor.size tensor should be considered.

From the example that @LaRiffle created:

` tensor = torch.rand(5, 3).send(bob)`
Calling `tensor.size()` would return:
--> `torch.size([5, 3])`
Instead, by calling tensor.size(dim), we can pass an arbitrary dimension that we would like to be returned, for instance:
Calling `tensor.size(0) `would return:
5`


`

**Comment by DanyEle** [comment]
Trying this code in an actual environment after installing the code in the branch "lstm", but I get the following  error when trying to send a tensor pointer to a worker.

image


The weird thing is that the corresponding test in test_hook.py is actually passing.

**Comment by LaRiffle** [comment]
As discussed with @DanyEle this is due to error while merging the code, this branch is broken after https://github.com/OpenMined/PySyft/pull/1995/commits/dc849169cfc18aa1b1c70c76fe3d9f7676f69a66

**Comment by DanyEle** [comment]
Appending these two lines at the end of the hook.py file adds support for LSTM/GRU cells and GRUs:

```
 #Daniele: added GRUs
        self.torch.nn.modules.rnn._rnn_impls["GRU"] = self.torch.gru

        ##Daniele:override _VF.LSTM_Cell and _VF.GRU_Cell with torch.LSTM_Cell and torch.GRU_Cell
        self.torch.nn.modules.rnn._VF = self.torch
```
